File: pi/shooter_tuner.py
"""
Shooter Parameter Tuner  —  Team 539
=====================================
Runs as a module inside main.py.  On startup it loads saved parameters from
/home/pi/shooter_tuner.json (creating the file with defaults if it does not
exist yet), publishes them to the ShooterTuner NetworkTable, and saves them
back whenever the robot code sends a save command.

This lets engineers iterate on shooter parameters from Elastic or Shuffleboard
during practice without recompiling the Java code.

NetworkTables layout  (table: "ShooterTuner")
----------------------------------------------
  Params/NearShotOffsetPercent    double  — percent offset applied to flywheel speed for near shots
  Params/FarShotOffsetPercent     double  — percent offset applied to flywheel speed for far shots
  Params/ReadyTolerance           double  — ft/s tolerance band around target speed to consider shooter ready
  Params/ShootReadyTimeoutSeconds double  — seconds to wait for shooter to reach ready speed before giving up
  Params/FloorFeedDelaySeconds    double  — seconds to delay floor roller engagement after feeder starts
  Params/Kp                       double  — flywheel proportional gain (written by ShooterAutoTuner)
  Params/FlywheelEfficiency       double  — ball-to-flywheel velocity ratio (0–1); tune to match field data
  Params/DragCoefficient          double  — aero drag constant B = 0.5×Cd×rho×A (kg/m); 0 disables drag
  Params/BallMassKg               double  — ball mass (kg); used in drag deceleration term B/m
  Cmd/Save                        boolean — Java sets True → Pi saves current values + resets
  Status/Heartbeat                integer — increments every loop for staleness detection
  Status/SavedOk                  boolean — True after the most recent save succeeded
  Status/ConfigFile               string  — absolute path of the active config file on the Pi

Persistent storage
-------------------
/home/pi/shooter_tuner.json  — written on first boot (defaults) and on every
                               save command.  Survives Pi reboots.
"""

import json
import os
import time
import ntcore

CONFIG_FILE = "/home/pi/shooter_tuner.json"

# ── Default parameter values (mirrors Java Constants.ShooterConstants /
#                              Constants.UltraShooterConstants) ──────────────
DEFAULTS: dict[str, float] = {
    "CloseShotOffsetPercent":   0.0,     # parabolic anchor at 1 m
    "MidShotOffsetPercent":     0.0,     # parabolic anchor at 4 m
    "FarShotOffsetPercent":     0.0,     # parabolic anchor at 7 m
    "ReadyTolerance":           1.745,   # ft/s — 100 RPM converted
    "ShootReadyTimeoutSeconds": 1.33,
    "FloorFeedDelaySeconds":    0.25,
    "Kp":                       0.003,   # flywheel proportional gain (autotuned)
    "FlywheelEfficiency":       0.49,    # ball exit speed / flywheel surface speed (0–1); calibrated vs ShooterOrca 2–5 m
    "DragCoefficient":          0.0132,  # B = 0.5×Cd×rho×A (kg/m); 0 = disable drag
    "BallMassKg":               0.270,   # ball mass (kg) for drag term B/m
}


# ── Helpers ────────────────────────────────────────────────────────────────────

def _load_or_create_config() -> dict[str, float]:
    """
    Loads the config from disk if it exists, filling any missing keys with
    defaults.  If no file exists, writes the defaults to disk immediately so
    permanent storage is established on first boot.
    """
    config = dict(DEFAULTS)

    if os.path.exists(CONFIG_FILE):
        try:
            with open(CONFIG_FILE, "r") as f:
                saved = json.load(f)
            # Accept only keys that are in DEFAULTS; fill gaps with defaults.
            for key in DEFAULTS:
                if key in saved:
                    config[key] = float(saved[key])
            print(f"[ShooterTuner] Loaded config from {CONFIG_FILE}")
        except Exception as exc:
            print(f"[ShooterTuner] Failed to load config ({exc}), using defaults")
    else:
        print(f"[ShooterTuner] No config file found — writing defaults to {CONFIG_FILE}")
        _save_config(config)  # create on first boot

    return config


def _save_config(config: dict[str, float]) -> bool:
    """Writes *config* to disk as pretty-printed JSON.  Returns True on success."""
    try:
        os.makedirs(os.path.dirname(CONFIG_FILE), exist_ok=True)
        with open(CONFIG_FILE, "w") as f:
            json.dump(config, f, indent=2)
        print(f"[ShooterTuner] Saved config to {CONFIG_FILE}")
        return True
    except Exception as exc:
        print(f"[ShooterTuner] Save failed: {exc}")
        return False


# ── Main entry point ───────────────────────────────────────────────────────────

def run(table: ntcore.NetworkTable) -> None:
    """
    Blocking loop.  Called from main.py in a daemon thread.

    :param table: Pre-initialized NT table (e.g. inst.getTable("ShooterTuner")).
    """
    params_table = table.getSubTable("Params")
    cmd_table    = table.getSubTable("Cmd")
    status_table = table.getSubTable("Status")

    # ── Load persistent config ─────────────────────────────────────────────────
    config = _load_or_create_config()

    # ── NT publishers for each parameter ──────────────────────────────────────
    param_pubs: dict[str, ntcore.DoublePublisher] = {}
    for key in DEFAULTS:
        pub = params_table.getDoubleTopic(key).publish()
        pub.set(config[key])
        param_pubs[key] = pub

    # ── NT subscribers — read values changed by the dashboard ─────────────────
    param_subs: dict[str, ntcore.DoubleSubscriber] = {}
    for key in DEFAULTS:
        param_subs[key] = params_table.getDoubleTopic(key).subscribe(config[key])

    # ── Save command ───────────────────────────────────────────────────────────
    # Java writes True to trigger a save; Pi resets it to False after handling.
    save_sub = cmd_table.getBooleanTopic("Save").subscribe(False)
    save_pub = cmd_table.getBooleanTopic("Save").publish()
    save_pub.set(False)

    # ── Status entries ─────────────────────────────────────────────────────────
    hb_pub        = status_table.getIntegerTopic("Heartbeat").publish()
    saved_ok_pub  = status_table.getBooleanTopic("SavedOk").publish()
    cfg_file_pub  = status_table.getStringTopic("ConfigFile").publish()

    cfg_file_pub.set(CONFIG_FILE)
    saved_ok_pub.set(os.path.exists(CONFIG_FILE))

    print("[ShooterTuner] Running — listening for save commands")

    heartbeat = 0
    while True:
        heartbeat += 1
        hb_pub.set(heartbeat)

        # ── Handle save command ────────────────────────────────────────────────
        if save_sub.get():
            # Snapshot the current NT values (may differ from what was loaded).
            for key in DEFAULTS:
                config[key] = param_subs[key].get()
            ok = _save_config(config)
            saved_ok_pub.set(ok)
            save_pub.set(False)  # acknowledge by resetting the flag

        time.sleep(0.02)  # 50 Hz — same cadence as other Pi modules

File: pi-java/src/main/java/ShooterTuner.java
/**
 * Shooter Parameter Tuner — Team 539
 *
 * <p>Mirrors pi/shooter_tuner.py exactly. Loads saved parameters from
 * {@code /home/pi/shooter_tuner.json} on startup (writing defaults if absent),
 * publishes them to the ShooterTuner NT table, and saves whenever the robot
 * sends a save command via {@code ShooterTuner/Cmd/Save = true}.
 *
 * <p>NT table: "ShooterTuner"
 * <ul>
 *   <li>{@code Params/*}          — one Double entry per tunable parameter</li>
 *   <li>{@code Cmd/Save}          — Java sets true → Pi saves + resets to false</li>
 *   <li>{@code Status/Heartbeat}  — increments every loop</li>
 *   <li>{@code Status/SavedOk}    — true after a successful save</li>
 *   <li>{@code Status/ConfigFile} — absolute path of the active config file</li>
 * </ul>
 */
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import edu.wpi.first.networktables.BooleanPublisher;
import edu.wpi.first.networktables.BooleanSubscriber;
import edu.wpi.first.networktables.DoublePublisher;
import edu.wpi.first.networktables.DoubleSubscriber;
import edu.wpi.first.networktables.IntegerPublisher;
import edu.wpi.first.networktables.NetworkTable;
import edu.wpi.first.networktables.StringPublisher;

import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.lang.reflect.Type;
import java.util.LinkedHashMap;
import java.util.Map;

public class ShooterTuner {

    private static final String CONFIG_FILE   = "/home/pi/shooter_tuner.json";
    private static final double LOOP_PERIOD_S = 0.02; // 50 Hz

    /**
     * Default values — mirror Java Constants.ShooterConstants /
     * Constants.UltraShooterConstants.
     */
    private static final Map<String, Double> DEFAULTS = new LinkedHashMap<>();
    static {
        DEFAULTS.put("CloseShotOffsetPercent",   0.0);
        DEFAULTS.put("MidShotOffsetPercent",     0.0);
        DEFAULTS.put("FarShotOffsetPercent",     0.0);
        DEFAULTS.put("ReadyTolerance",           1.745);  // ft/s ≈ 100 RPM
        DEFAULTS.put("ShootReadyTimeoutSeconds", 1.33);
        DEFAULTS.put("ShootWaitSeconds",         0.5);
        DEFAULTS.put("FloorFeedDelaySeconds",    0.25);
        DEFAULTS.put("Kp",                       0.003);
        DEFAULTS.put("FlywheelEfficiency",       0.43);
        DEFAULTS.put("DragCoefficient",          0.0132);
        DEFAULTS.put("BallMassLbs",              0.595);
    }

    private static final Gson GSON     = new GsonBuilder().setPrettyPrinting().create();
    private static final Type MAP_TYPE = new TypeToken<Map<String, Double>>(){}.getType();

    // ── Config helpers ────────────────────────────────────────────────────────────

    private static Map<String, Double> loadOrCreate() {
        Map<String, Double> config = new LinkedHashMap<>(DEFAULTS);
        File f = new File(CONFIG_FILE);

        if (f.exists()) {
            try (FileReader r = new FileReader(f)) {
                Map<String, Double> saved = GSON.fromJson(r, MAP_TYPE);
                if (saved != null) {
                    for (String key : DEFAULTS.keySet()) {
                        if (saved.containsKey(key)) config.put(key, saved.get(key));
                    }
                }
                System.out.println("[ShooterTuner] Loaded config from " + CONFIG_FILE);
            } catch (Exception e) {
                System.out.println("[ShooterTuner] Failed to load config (" + e.getMessage() + "), using defaults");
            }
        } else {
            System.out.println("[ShooterTuner] No config file found — writing defaults to " + CONFIG_FILE);
            save(config);
        }
        return config;
    }

    private static boolean save(Map<String, Double> config) {
        try {
            new File(CONFIG_FILE).getParentFile().mkdirs();
            try (FileWriter w = new FileWriter(CONFIG_FILE)) {
                GSON.toJson(config, w);
            }
            System.out.println("[ShooterTuner] Saved config to " + CONFIG_FILE);
            return true;
        } catch (Exception e) {
            System.out.println("[ShooterTuner] Save failed: " + e.getMessage());
            return false;
        }
    }

    // ── NT loop ───────────────────────────────────────────────────────────────────

    /**
     * Blocking 50 Hz loop. Called from {@link Main} in a daemon thread.
     *
     * @param table pre-initialized "ShooterTuner" NT table
     */
    public static void run(NetworkTable table) {
        NetworkTable paramsTable = table.getSubTable("Params");
        NetworkTable cmdTable    = table.getSubTable("Cmd");
        NetworkTable statusTable = table.getSubTable("Status");

        Map<String, Double> config = loadOrCreate();

        // Publish initial values and open subscribers for dashboard edits.
        Map<String, DoublePublisher>  pubs = new LinkedHashMap<>();
        Map<String, DoubleSubscriber> subs = new LinkedHashMap<>();

        for (Map.Entry<String, Double> e : DEFAULTS.entrySet()) {
            DoublePublisher pub = paramsTable.getDoubleTopic(e.getKey()).publish();
            pub.set(config.get(e.getKey()));
            pubs.put(e.getKey(), pub);
            subs.put(e.getKey(), paramsTable.getDoubleTopic(e.getKey()).subscribe(config.get(e.getKey())));
        }

        // Save command: Java writes true → Pi saves and resets to false.
        BooleanSubscriber saveSub = cmdTable.getBooleanTopic("Save").subscribe(false);
        BooleanPublisher  savePub = cmdTable.getBooleanTopic("Save").publish();
        savePub.set(false);

        // Status entries.
        IntegerPublisher hbPub      = statusTable.getIntegerTopic("Heartbeat").publish();
        BooleanPublisher savedOkPub = statusTable.getBooleanTopic("SavedOk").publish();
        StringPublisher  cfgFilePub = statusTable.getStringTopic("ConfigFile").publish();

        cfgFilePub.set(CONFIG_FILE);
        savedOkPub.set(new File(CONFIG_FILE).exists());

        System.out.println("[ShooterTuner] Running — listening for save commands.");

        long heartbeat = 0;
        long loopNs    = (long) (LOOP_PERIOD_S * 1_000_000_000L);

        while (true) {
            long start = System.nanoTime();

            hbPub.set(heartbeat++);

            if (saveSub.get()) {
                for (String key : DEFAULTS.keySet()) {
                    config.put(key, subs.get(key).get());
                }
                savedOkPub.set(save(config));
                savePub.set(false);
            }

            long sleep = loopNs - (System.nanoTime() - start);
            if (sleep > 0) {
                try { Thread.sleep(sleep / 1_000_000L, (int) (sleep % 1_000_000L)); }
                catch (InterruptedException e) { Thread.currentThread().interrupt(); return; }
            }
        }
    }
}
